python/activator/middleware_interface.py

```diff
@@ -492,24 +578,128 @@ def run_pipeline(self, visit: Visit, exposure_ids: set) -> None:
+        """Return the set of dataset types that can be safely merged from a worker.
+
+        Parameters
+        ----------
+        butler : `lsst.daf.butler.Butler`
+            The butler in which to search for dataset types.
+
+        Returns
+        -------
+        types : iterable [`str`]
+            The dataset types to return.
+        """
+        return [dstype for dstype in butler.registry.queryDatasetTypes(...)
+                if "detector" in dstype.dimensions]
+
+    def _export_subset(self, visit: Visit, exposure_ids: set[int],
+                       dataset_types: typing.Any, in_collections: typing.Any, out_collection: str) -> None:
+        """Copy datasets associated with a processing run back to the
+        central Butler.
+
+        Parameters
+        ----------
+        visit : Visit
+            The visit whose outputs need to be exported.
+        exposure_ids : `set` [`int`]
+            Identifiers of the exposures that were processed.
+        dataset_types
+            The dataset type(s) to transfer; can be any expression described in
+            :ref:`daf_butler_dataset_type_expressions`.
+        in_collections
+            The collections to transfer from; can be any expression described
+            in :ref:`daf_butler_collection_expressions`.
+        out_collection : `str`
+            The chained collection in which to include the datasets. Need not
+            exist before the call.
+        """
+        try:
+            # Need to iterate over datasets at least twice, so list.
+            datasets = list(self.butler.registry.queryDatasets(
+                dataset_types,
+                collections=in_collections,
+                # in_collections may include other runs, so need to filter.
+                # Since AP processing is strictly visit-detector, these three
+                # dimensions should suffice.
+                # DO NOT assume that visit == exposure!
+                where=f"exposure in ({', '.join(str(x) for x in exposure_ids)})",
+                instrument=self.instrument.getName(),
+                detector=visit.detector,
+            ))
+        except lsst.daf.butler.registry.DataIdError as e:
+            raise ValueError("Invalid visit or exposures.") from e
+        if not datasets:
+            raise ValueError(f"No datasets match visit={visit} and exposures={exposure_ids}.")
+
+        with tempfile.NamedTemporaryFile(mode="w+b", suffix=".yaml") as export_file:
+            # TODO: saveDatasets(elements={"exposure", "visit"}) doesn't work.
+            # Use import(skip_dimensions) until DM-36062 is fixed.
+            with self.butler.export(filename=export_file.name) as export:
+                export.saveDatasets(datasets)
+            self.central_butler.import_(filename=export_file.name,
+                                        directory=self.butler.datastore.root,
+                                        skip_dimensions={"instrument", "detector",
```

Comment:
I'm not sure why `skip_dimensions` is needed.  Is this just for efficiency?

Reply:
IIRC, `import` has problems merging dimensions, especially governor dimensions. As noted in the comment above, we shouldn't need to redefine anything other than `exposure` and `visit`, but there's an as yet unfixed bug that prevents us from expressing that directly.

Reply:
Ah, now I remember what it was: merging governor dimensions can trigger a deadlock in the registry. Excluding the other dimensions was just for caution's sake.